book/src/dev/rfcs/0005-treestate.md

```diff
@@ -0,0 +1,194 @@
+been satisfied and the Spend description can be validated independently. For
+JoinSplit's, if it's not a previously finalized block anchor, it must be the
+treestate anchor of previous JoinSplit in this transaction, and we have to wait
+for that one to be parsed and its root computed to check that ours is
+valid. Luckily, it can only be a previous JoinSplit in this transaction, and are
+[usually the immediately previous one](zcashd), so the set of candidate anchors
+is smaller for earlier JoinSplit's in a transcation, but larger for the later
+ones. For these JoinSplit's, they can be validated independently of their
+anchor's finalization status as long as the final check of the anchor is done,
+when available, such as at the Transaction level after all the JoinSplit's have
+finished validating everything that can be validated without the context of
+their anchor's finalization state.
+
+So for each transaction, for both Spend descriptions and JoinSplit's, we can
+pre-emptively try to do our consensus check by looking up the anchors in our
+finalized that first. For Spend's, we then trigger the remaining validation and
+when that finishes we are full done with those. For JoinSplit's, the anchor
+state check may pass early if it's a previous block Sprout note commitment tree
+root, but it may fail because it's an earlier JoinSplit's root instead, so once
+the JoinSplit validates independently of the anchor, we wait for all candidate
+previous JoinSplit's in that transaction finish validating before doing the
+anchor consensus check again, but against the output treestate roots of earlier
+JoinSplit's.
+
+Both Sprout and Sapling note commitment trees must be computed for the whole
+block to validate, but Sprout, we need to compute interstitial treestates in
+between JoinSplit's in order to do the final consensus check for each/all
+JoinSplit's, not just for the whole block, like in Sapling.
+
+For Sapling, at the block layer, we can iterate over all the transactions in
+order and if they have Spend's and/or Output's, we update our Nullifer set for
+the block as nullifiers are revealed in Spend descriptions, and update our note
+commitment tree as note commitments are revealed in Output descriptions, adding
+them as leaves in positions according to their order as they appear transaction
+to transction, output to output, in the block. This can be done independent of
+the transaction validations. When the Sapling transactions are all validated,
+the note commitment tree root should be computed: this is the anchor for this
+block. For Sapling and Blossom blocks, we need to check that this root matches
+the `RootHash` bytes in this block's header, as the `FinalSaplingRoot`. Once all
+other consensus and validation checks are done, this will be saved down to our
+finalized state to our `sapling_anchors` set, making it available for lookup by
+other Sapling descriptions in future transactions.
```

Comment:
Should we talk about Heartwood and Canopy here too?
So we explain why they aren't included?

Reply:
Added a TODO for this for now: https://github.com/ZcashFoundation/zebra/pull/983/commits/997efea18fb32124b3e05dca2f5670cd32f6af6b